private/bufpkg/bufplugin/bufpluginconfig/config.go

```diff
@@ -0,0 +1,191 @@
+		return nil, nil
+	}
+	if isArchiveEmpty && isGoEmpty && !isNPMEmpty {
+		npmRuntimeConfig, err := newNPMRuntimeConfig(externalRuntimeConfig.NPM)
+		if err != nil {
+			return nil, err
+		}
+		return &RuntimeConfig{
+			NPM: npmRuntimeConfig,
+		}, nil
+	}
+	if isArchiveEmpty && !isGoEmpty && isNPMEmpty {
+		goRuntimeConfig, err := newGoRuntimeConfig(externalRuntimeConfig.Go)
+		if err != nil {
+			return nil, err
+		}
+		return &RuntimeConfig{
+			Go: goRuntimeConfig,
+		}, nil
+	}
+	if !isArchiveEmpty && isGoEmpty && isNPMEmpty {
+		archiveRuntimeConfig, err := newArchiveRuntimeConfig(externalRuntimeConfig.Archive)
+		if err != nil {
+			return nil, err
+		}
+		return &RuntimeConfig{
+			Archive: archiveRuntimeConfig,
+		}, nil
+	}
+	// If we made it this far, that means the config specifies multiple
+	// runtime languages.
+	//
+	// We might eventually want to support multiple runtime configuration
+	// (e.g. 'go' and 'archive'), but it's safe to start with an error for
+	// now.
+	return nil, fmt.Errorf("%s configuration contains multiple runtime languages", ExternalConfigFilePath)
+}
+
+func newNPMRuntimeConfig(externalNPMRuntimeConfig ExternalNPMRuntimeConfig) (*NPMRuntimeConfig, error) {
+	if err := validateRuntimeDeps(externalNPMRuntimeConfig.Deps); err != nil {
+		return nil, err
+	}
+	return &NPMRuntimeConfig{
+		Deps: externalNPMRuntimeConfig.Deps,
+	}, nil
+}
+
+func newGoRuntimeConfig(externalGoRuntimeConfig ExternalGoRuntimeConfig) (*GoRuntimeConfig, error) {
+	if err := validateRuntimeDeps(externalGoRuntimeConfig.Deps); err != nil {
+		return nil, err
+	}
+	// The best we can do is verify that the minimum version
+	// is a valid semantic version, just like we do for the
+	// runtime dependencies.
+	//
+	// This will not actually verify that the go version is
+	// in the valid set. It's impossible to capture the
+	// real set of valid identifiers at any given time (for
+	// an old version of the buf CLI) without reaching out to
+	// some external source at runtime.
+	//
+	// Note that this ensures the user's configuration specifies
+	// a 'v' prefix in the version (e.g. v1.18) even though the
+	// minimum version is rendered without it in the go.mod.
+	if externalGoRuntimeConfig.MinVersion != "" && !semver.IsValid(externalGoRuntimeConfig.MinVersion) {
+		return nil, fmt.Errorf("the go minimum version %q must be a valid semantic version", externalGoRuntimeConfig.MinVersion)
+	}
+	return &GoRuntimeConfig{
+		MinVersion: externalGoRuntimeConfig.MinVersion,
+		Deps:       externalGoRuntimeConfig.Deps,
+	}, nil
+}
+
+func newArchiveRuntimeConfig(externalArchiveRuntimeConfig ExternalArchiveRuntimeConfig) (*ArchiveRuntimeConfig, error) {
+	if err := validateRuntimeDeps(externalArchiveRuntimeConfig.Deps); err != nil {
+		return nil, err
+	}
+	return &ArchiveRuntimeConfig{
+		Deps: externalArchiveRuntimeConfig.Deps,
+	}, nil
+}
+
+func validateRuntimeDeps(dependencies []string) error {
```

Comment:
Instead of defining a format, since we already have language specific config, I think we should have language specific dependency definitions. It gives us,
- Idiomatic names (modules for go, packages for npm etc...)
- Language specific constraints (simple semver for go, semver with constraints for npm)
- Won't have to worry about archive registy as they can include anything that is at least human readable (package/version)

Reply:
In general I agree, but I don't like that it puts more cognitive overhead on the user to understand how to specify dependencies for the specific runtime in question - it'd be great if we had standard way of specifying `deps` for the user-facing configuration (what this attempts to propose). This was something I discussed with @bufdev.

With that said, I admit that it feels like this is susceptible to break down at some point - we just need to be mindful of the ergonomics here.